Add EKS + OIDC for per Pod IAM blog post (#3232)
pulumi/docs: content/blog/eks-oidc/index.md

---
date: "2020-05-08"

> title: "Access Control for Pods on AWS EKS"

Reviewer: :-) https://aws.amazon.com/eks/
```suggestion
title: "Access Control for Pods on Amazon EKS"
```

authors: ["mike-metral"]
tags: ["EKS", "RBAC", "Kubernetes"]
meta_desc: "AWS EKS clusters can use IAM roles and policies for Pods to assign fine-grained access control of AWS services."
meta_image: cluster.svg
---

> AWS [EKS][aws-eks] clusters can use [IAM][aws-iam] roles and policies for Pods

Reviewer: ```suggestion
Amazon [EKS][aws-eks] clusters can use [IAM][aws-iam] roles and policies for Pods
```

to assign fine-grained access control of AWS services. The AWS IAM entities map into Kubernetes
[RBAC][k8s-rbac] to configure the permissions of Pods that work with AWS
services.

Together, AWS IAM and Kubernetes RBAC enable least-privileged access for your
apps, scoped to the appropriate policies and user requirements.

[k8s-rbac]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/

<!--more-->

## Overview

* [Pod Access Control](#pod-access-control)
* [Create an OIDC provider](#create-an-oidc-provider)
* [Create IAM for a S3 app](#create-iam-for-a-s3-app)
* [Deploy a S3 app](#deploy-a-s3-app)
* [Wrap-Up](#wrap-up)
* [Next Steps](#next-steps)

## Pod Access Control

AWS EKS supports using IAM entities in a Pod [Service Account][k8s-sa] by
leveraging an [OIDC provider][aws-oidc] connected to the Kubernetes cluster.

> Continuing with the example from the [AWS blog post][aws-pod-iam], when an [S3 app written in Go][s3-app] pushes

Reviewer: Perhaps call out what this blog post is about, and how it relates to the post here?  Related - maybe highlight what this post is going to do as part of the overview above?  Perhaps using generally the same sentence as in the wrap-up?

> In <100 lines of code we've shown how EKS Pods can use AWS IAM to create fine-grained
permissions for apps that integrate with other AWS services.

Author: Done

an object to a bucket with the AWS SDK, it will need write access to S3.

When a Pod is launched with a particular Service Account, the OIDC provider works
with Kubernetes to verify the Pod's identity, and in turn collaborates with the
[AWS Secure Token Service (STS)][aws-sts] to grant the Pod temporary
credentials to use with the IAM role.

## Create an OIDC provider

Creating an OIDC provider is as simple as toggling the `createOidcProvider` option
in the definition of your EKS cluster.

When enabled, the OIDC provider will be created and associated with the
cluster's OIDC provider URL.

![Create an EKS Cluster and OIDC provider](cluster.svg)

## Create IAM for a S3 app

We'll use the OIDC provider URL and Amazon Resource Name (ARN) to compose the
[AssumeRoleWithWebIdentity][aws-assume-role-web], and S3 IAM policies that will
be attached to a new S3 IAM role.

After the role is configured, a Service Account for the S3 Pod will be
created, and annotated with the ARN of the S3 role to bind the two together.

> ![Create the AWS IAM and Kubernetes Service Account for the Pod](iam.svg)

Author: The images were used to break up text-heavy content, and because they render quite nicely as SVGs, so it creates good visuals in between text.

Are you saying that instead of capturing the code samples they be in codeblocks for copy/paste?

> ![Create the AWS IAM and Kubernetes Service Account for the Pod](iam.svg)

Reviewer 2: Yeah, that's right. I don't have a strong opinion on it, but that's the feedback I've received on my own posts.

> ![Create the AWS IAM and Kubernetes Service Account for the Pod](iam.svg)

Author: cc @spara Thoughts on codeblocks vs SVGs?

> ![Create the AWS IAM and Kubernetes Service Account for the Pod](iam.svg)

Reviewer 3: checking how it renders

> ![Create the AWS IAM and Kubernetes Service Account for the Pod](iam.svg)

Reviewer 3: I think we should stick to code blocks for copy-paste reasons and readability. While the svgs are clear, the text renders at different sizes, which makes it difficult to read on a mobile device:

![Screen Shot 2020-05-19 at 2 36 31 PM](https://user-images.githubusercontent.com/638672/82370444-2c1b5680-99de-11ea-9acc-aad79d0d7904.png)

---
date: "2020-05-08"
title: "Access Control for Pods on Amazon EKS"
authors: ["mike-metral"]
tags: ["EKS", "RBAC", "Kubernetes"]
meta_desc: "Amazon EKS clusters can use IAM roles and policies for Pods to assign fine-grained access control of AWS services."
meta_image: cluster.svg
---

Amazon [EKS][aws-eks] clusters can use [IAM][aws-iam] roles and policies for Pods
to assign fine-grained access control of AWS services. The AWS IAM entities map into Kubernetes

> [RBAC][k8s-rbac] to configure the permissions of Pods that work with AWS

Reviewer 3: ```suggestion
[RBAC][k8s-rbac] to configure Pod's permissions that work with AWS
```

services.

Together, AWS IAM and Kubernetes RBAC enable least-privileged access for your
apps, scoped to the appropriate policies and user requirements.

[k8s-rbac]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/

<!--more-->

## Overview

In <100 lines of code we'll demonstrate how EKS Pods can use AWS IAM to create
fine-grained permissions for apps that integrate with other AWS services.

* [Pod Access Control](#pod-access-control)
* [Create an OIDC provider](#create-an-oidc-provider)
* [Create IAM for a S3 app](#create-iam-for-a-s3-app)
* [Deploy a S3 app](#deploy-a-s3-app)
* [Wrap-Up](#wrap-up)
* [Next Steps](#next-steps)

## Pod Access Control

AWS EKS supports using IAM entities in a Pod [Service Account][k8s-sa] by
leveraging an [OIDC provider][aws-oidc] connected to the Kubernetes cluster.

Continuing with the example from the [AWS blog post][aws-pod-iam], when an [S3 app written in Go][s3-app] pushes
an object to a bucket with the AWS SDK, it will need write access to S3.

When a Pod is launched with a particular Service Account, the OIDC provider works

> with Kubernetes to verify the Pod's identity, and in turn collaborates with the

Reviewer 3: ```suggestion
with Kubernetes to verify the Pod's identity, and in turn, collaborates with the
```

> with Kubernetes to verify the Pod's identity, and in turn collaborates with the

Reviewer 3: ```suggestion
with Kubernetes to verify the Pod's identity, and in turn, collaborates with the
```

[AWS Secure Token Service (STS)][aws-sts] to grant the Pod temporary
credentials to use with the IAM role.

## Create an OIDC provider

Creating an OIDC provider is as simple as toggling the `createOidcProvider` option
in the definition of your EKS cluster.

When enabled, the OIDC provider will be created and associated with the
cluster's OIDC provider URL.

![Create an EKS Cluster and OIDC provider](cluster.svg)

## Create IAM for a S3 app

We'll use the OIDC provider URL and Amazon Resource Name (ARN) to compose the

> [AssumeRoleWithWebIdentity][aws-assume-role-web], and S3 IAM policies that will

Reviewer 3: ```suggestion
[AssumeRoleWithWebIdentity][aws-assume-role-web] and S3 IAM policies that will
```

be attached to a new S3 IAM role.

> After the role is configured, a Service Account for the S3 Pod will be

Reviewer 3: ```suggestion
After the role is configured, a Service Account for the S3 Pod is
```

> created, and annotated with the ARN of the S3 role to bind the two together.

Reviewer 3: ```suggestion
created and annotated with the ARN of the S3 role to bind the two together.
```

![Create the AWS IAM and Kubernetes Service Account for the Pod](iam.svg)

## Deploy a S3 app

We'll deploy the [S3 app][peks-oidc] to use the new IAM-backed Service Account.

Once the Pod is running, the Service Account annotation will be automatically
managed by a [Kubernetes dynamic admission controller][k8s-dynamic-webhook] run
by EKS on your behalf.

> The [AWS EKS webhook][eks-webhook] manages Pod identity, and injects STS

Reviewer 3: ```suggestion
The [AWS EKS webhook][eks-webhook] manages Pod identity and injects STS
```

> The [AWS EKS webhook][eks-webhook] manages Pod identity, and injects STS

Reviewer 3: ```suggestion
The [AWS EKS webhook][eks-webhook] manages Pod identity and injects STS
```

managed by a [Kubernetes dynamic admission controller][k8s-dynamic-webhook] run
by EKS on your behalf.

The [AWS EKS webhook][eks-webhook] manages Pod identity, and injects STS
credentials into the Pod to use with the S3 role.

With the credentials, the app can successfully upload data to S3.

```typescript
import * as aws from "@pulumi/aws";
import * as k8s from "@pulumi/kubernetes";
import * as pulumi from "@pulumi/pulumi";

const bucket = new aws.s3.Bucket("pod-irsa-job-bucket");
const bucketName = bucket.id;
const regionName = pulumi.output(aws.getRegion({}, {async: true})).name;
const s3Pod = new k8s.core.v1.Pod(podName,
    {
        metadata: {labels: labels, namespace: appsNamespaceName},
        spec: {
            serviceAccountName: sa.metadata.name,
            containers: [
                {
                    name: podName,
                    image: "amazonlinux:2018.03",
                    command: ["sh", "-c",
                        pulumi.interpolate`curl -sL -o /s3-echoer https://git.io/JfnGX && chmod +x /s3-echoer && echo This is an in-cluster test | /s3-echoer ${bucketName} && sleep 3600`,
                    ],
                    env: [

>                         {name: "AWS_DEFAULT_REGION", value: pulumi.interpolate`${regionName}`},

Author: It does not work unfortunately. I tried this as-is before using interpolation and the Pod would error. Here's the `kubectl describe on the pod`:

```
      AWS_DEFAULT_REGION:           Calling [toString] on an [Output<T>] is not supported.

                                    To get the value of an Output<T> as an Output<string> consider either:
                                    1: o.apply(v => `prefix${v}suffix`)
                                    2: pulumi.interpolate `prefix${v}suffix`

                                    See https://pulumi.io/help/outputs for more details.
                                    This function may throw in a future version of @pulumi/pulumi.
      ENABLE_IRP:                   true
      AWS_ROLE_ARN:                 arn:aws:iam::153052954103:role/s3-0cd6cf6
      AWS_WEB_IDENTITY_TOKEN_FILE:  /var/run/secrets/eks.amazonaws.com/serviceaccount/token
```

cluster's OIDC provider URL.

```typescript
import * as eks from '@pulumi/eks'

// Create an EKS cluster with default settings.
// Create and attach an OIDC provider to the cluster.
const cluster = new eks.Cluster('myCluster', {
      createOidcProvider: true,
});
```

## Create IAM for a S3 app

We'll use the OIDC provider URL and Amazon Resource Name (ARN) to compose the
[AssumeRoleWithWebIdentity][aws-assume-role-web] and S3 IAM policies that will
be attached to a new S3 IAM role.

After the role is configured, a Service Account for the S3 Pod will be
created and annotated with the ARN of the S3 role to bind the two together.

```typescript
import * as aws from '@pulumi/aws'
import * as k8s from '@pulumi/kubernetes'
import * as pulumi from '@pulumi/pulumi'

// Create a pulumi Kubernetes provider using the cluster's kubeconfig.
const k8sProvider = new k8s.Provider('k8s', {
  kubeconfig: cluster.kubeconfig.apply(JSON.stringify),
});

// Create a k8s namespace in the cluster.
const namespace = new k8s.core.v1.Namespace('apps', undefined, { k8sProvider });

// Get the OIDC provider's URL for the cluster.
const clusterOidcProvider = cluster.core.oidcProvider.url;

> // Create the IAM AssumeRoleWebWebIdentity policy for the Service Account.

Author: Clarified this -- thank you.

  policyArn: 'arn:aws:iam::aws:policy/AmazonS3FullAccess',
  role: saRole,
});

// Create a Service Account with the IAM role annotated to use with the Pod.
const sa = new k8s.core.v1.ServiceAccount(
  saName,
  {
    metadata: {
      namespace: namespace.metadata.name,
      name: saName,
      annotations: {
        'eks.amazonaws.com/role-arn': saRole.arn,
      },
    },
  },
  { k8sProvider });
```

## Deploy a S3 app

We'll deploy the [S3 app][peks-oidc] to use the new IAM-backed Service Account.

Once the Pod is running, the Service Account annotation will be automatically
managed by a [Kubernetes dynamic admission controller][k8s-dynamic-webhook] run
by EKS on your behalf.

The [AWS EKS webhook][eks-webhook] manages Pod identity and injects STS
credentials into the Pod to use with the S3 role.

With the credentials, the app can successfully upload data to S3.

```typescript
import * as aws from "@pulumi/aws";
import * as k8s from "@pulumi/kubernetes";
import * as pulumi from "@pulumi/pulumi";

const bucket = new aws.s3.Bucket("pod-irsa-job-bucket");
const bucketName = bucket.id;
const regionName = pulumi.output(aws.getRegion({}, {async: true})).name;

> const s3Pod = pulumi.all([bucketName, regionName]).apply(([bName, region]) => {

Author: Corrected. Thanks!
